File: solutions/csharp/sieve/Sieve.cs
using System;
using System.Collections;
using System.Collections.Generic;

public static class Sieve
{
    public static IEnumerable<int> Primes(int limit)
    {
        if (limit < 0) throw new ArgumentOutOfRangeException(nameof(limit));

        var primes = new BitArray(limit + 1);
        for (var i = 2; i <= limit; i++)
        {
            if (primes[i]) continue;
            for (var j = i * 2; j <= limit; j += i)
            {
                primes[j] = true;
            }
            yield return i;
        }
    }
}

File: solutions/ruby/sieve/sieve.rb
class Sieve
  def initialize(limit)
    @limit = limit
  end

  def primes
    candidates = [nil, nil, *2..@limit]
    (2..Math.sqrt(@limit)).each do |n|
      next unless candidates[n]
      (n * n..@limit).step(n) { |m| candidates[m] = nil }
    end
    candidates.compact
  end
end
